agency/agency_connection.go

```diff
@@ -0,0 +1,295 @@
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+
+	if len(c.connections) == 0 {
+		return nil, driver.WithStack(fmt.Errorf("no connections"))
+	}
+	r, err := c.connections[0].NewRequest(method, path)
+	if err != nil {
+		return nil, driver.WithStack(err)
+	}
+	return r, nil
+}
+
+// Do performs a given request, returning its response.
+// In case of a termporary failure, the request is retried until
+// the deadline is exceeded.
+func (c *agencyConnection) Do(ctx context.Context, req driver.Request) (driver.Response, error) {
+	if ctx == nil {
+		ctx = context.Background()
+	}
+	deadline, ok := ctx.Deadline()
+	if !ok {
+		deadline = time.Now().Add(time.Second * 30)
+	}
+	timeout := time.Until(deadline)
+	if timeout < minAgencyTimeout {
+		timeout = minAgencyTimeout
+	}
+	attempt := 1
+	delay := agencyConnectionFailureBackoff(0)
+	for {
+		lctx, cancel := context.WithTimeout(ctx, timeout/3)
+		resp, isPerm, err := c.doOnce(lctx, req)
+		cancel()
+		if err == nil {
+			// Success
+			return resp, nil
+		} else if isPerm {
+			// Permanent error
+			return nil, driver.WithStack(err)
+		}
+		// Is deadline exceeded?
+		if time.Now().After(deadline) {
+			return nil, driver.WithStack(fmt.Errorf("All %d attemps resulted in temporary failure", attempt))
+		}
+		// Just retry
+		attempt++
+		delay = agencyConnectionFailureBackoff(delay)
+		// Wait a bit so we don't hammer the agency
+		select {
+		case <-time.After(delay):
+			// Continue
+		case <-ctx.Done():
+			// Context canceled
+			return nil, driver.WithStack(ctx.Err())
+		}
+	}
+}
+
+// Do performs a given request once, returning its response.
+// Returns: Response, isPermanentError, Error
+func (c *agencyConnection) doOnce(ctx context.Context, req driver.Request) (driver.Response, bool, error) {
+	c.mutex.RLock()
+	connections := c.connections
+	c.mutex.RUnlock()
+
+	if len(c.connections) == 0 {
+		return nil, true, driver.WithStack(fmt.Errorf("no connections"))
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	results := make(chan driver.Response, len(connections))
+	errors := make(chan error, len(connections))
+	wg := sync.WaitGroup{}
+	for _, epConn := range connections {
+		wg.Add(1)
+		go func(epConn driver.Connection) {
+			defer wg.Done()
+			epReq := req.Clone()
+			result, err := epConn.Do(ctx, epReq)
+			if err == nil {
+				// Success
+				results <- result
+				// Cancel all other requests
+				cancel()
+				return
+			}
+			// Check error
+			if statusCode, ok := isArangoError(err); ok {
+				// We have a status code, check it
+				if statusCode >= 400 && statusCode < 500 && statusCode != 408 {
+					// Permanent error, return it
+					errors <- driver.WithStack(err)
+					// Cancel all other requests
+					cancel()
+					return
+				}
+			}
+			// No permanent error, try next agent
+		}(epConn)
+	}
+
+	// Wait for go routines to finished
+	wg.Wait()
+	cancel()
+	close(results)
+	close(errors)
+	if result, ok := <-results; ok {
```

Comment:
I do not understand what happens if a 307 is returned. Is this reported as an error? Or does each request follow the redirect? Both would be bad. I think if we send to all agents we need to ignore 307, but I do not understand how this works here.

Reply:
The underling HTTP connection is configured to fail in redirect. That means a 302 error is returned. The code here will consider any non-permanent failure (everything in the 300 range) as "this agent did not handle, so we'll wait for another".

Only in case all agents did not respond with either a result or a permanent failure, we'll retry the operation.
So the notion of automatic retry does not exist in any other case.

Reply:
Good, understood, thanks.